### DISCUSSION TITLE
8 channel output (e.g., HiFiBerry DAC8x) [DONE]

### DISCUSSION THREAD

**Discussion by SetTheTone** [probonopd/MiniDexed#655] [Ideas]
I'm a bit new here both on GitHub and MiniDexed so don't give me a hard time, I've searched and cannot find anything so here goes. I've noticed that HiFiBerry now have an 8 DAC output card DAC8x would it be possible to use this with MiniDexed for separate outs for each Dexed? See Link Below

[https://www.hifiberry.com/blog/more-than-just-stereo-the-dac8x/](url)

**Comment by diyelectromusic** [comment]
Not without changing the code to send the outputs of each TG to an independent DAC.

It looks like it is somehow enabling the new multiple I2S channels, so that would have to be supported in the base environment that MiniDexed uses (circle) before we could update the MiniDexed code itself.  I think at the moment circle only supports two-channel I2S - presumably input and output.

It would be a really interesting thing to contemplate, but at this point is probably a fair bit of work and would need a circle update...

Kevin

**Comment by diyelectromusic** [comment]
Although it looks like the dac8x is essentially four PCM5102s...(*) so now that has me wondering what it would take...

* See https://github.com/raspberrypi/linux/blob/55faf2eb5190410941aaabdc5bbd2c74c27f546e/sound/soc/bcm/rpi-simple-soundcard.c#L322

It appears to be using GPIO 18, 19, 21, 23, 25, 27 - for CLK, WD, then DO1 to DO4... (more here: https://github.com/raspberrypi/linux/blob/55faf2eb5190410941aaabdc5bbd2c74c27f546e/arch/arm/boot/dts/overlays/hifiberry-dac8x-overlay.dts#L13)

I'm not sure where the 8 comes from - presumably that is four stereo channels...?

So we'd have to make each TG mono to do that and assume that all panning and mixing was done externally.  It does however suggest that one could do the same thing with four GY-PCM5102 modules...

Kevin

**Reply by SetTheTone** [reply]
Hi Kevin

Thanks for your reply, I'm not as techy as you, I'm just good at following the idiot guides you good people write.

As I understand the TX816 was mono out per card, it would not be bad to sacrifice the stereo, as it can created in a DAW or analogue mixing desk.

Hopefully, my post has sparked something to enhance the MiniDexed, not that it is required but it would be a lovely fork to this excellent synth.

Milton

**Reply by diyelectromusic** [reply]
It looks like Rene is going to see what it would take to pop this into circle (https://github.com/rsta2/circle/discussions/453), then I could take a look at how MiniDexed might use it.

Of course, I'll have to try with four GY-PCM5102As as I don't have a dac8x :)

Kevin

**Reply by diyelectromusic** [reply]
Well if Rene does manage to get this updated then I might send this off to be built and give it a go...

![RPi-QuadDAC-MiniDexed-3d](https://github.com/probonopd/MiniDexed/assets/68612569/3149e317-8fda-4191-b48f-eaea269cdc45)

**Reply by diyelectromusic** [reply]
Ok, so four PCM5102s work in Raspberry Pi OS (see https://diyelectromusic.wordpress.com/2024/05/27/rpi-5-quad-stereo-sound-with-pcm5102a/) and I've now managed to build circle with them too (Rene added support yesterday) and have them working.

I'll have a look and see what would be involved in getting MiniDexed to use them :)

Kevin

**Reply by Kirtai** [reply]
Very cool. I might get a Pi5 just for this and the higher polyphony :)

**Reply by diyelectromusic** [reply]
That reminds me - I still need to look at the polyphony question for both Pi 4 and 5...

**Reply by diyelectromusic** [reply]
Seems like it works :)

Details here: https://diyelectromusic.com/2024/06/09/minidexed-quad-dac-pcb-design/

![IMG_7861](https://github.com/probonopd/MiniDexed/assets/68612569/87538a95-0ba2-4a65-8eb6-1f0f8aad3c40)

**Comment by probonopd** [comment]
Would the purpose be to output each of the tone generators on its own channel? Then we'd need 8...

**Reply by SetTheTone** [reply]
Hi probonopd

Yes, but hard right and hard left each stereo pair of each pair of Dexed modules so they are mono something like 1L 2R to 7L 8R on the DAC board, you can ultimately make the stereo paired in a mixer or DAW if required, I see no reason why they should be tied to each other but correct me if I am wrong in thinking this.

Regards
Milton

**Reply by probonopd** [reply]
...or is what you _really_ would like to see a MiniDexed VST for the DAW? :)

**Reply by diyelectromusic** [reply]
I must admit a certain sympathy to supporting such things in hardware.  Today just about anything can be done in a DAW, but I have to say there is almost too much choice a very few barriers.  I do think that often the fun (and creativity) comes from working within the constraints of actual devices, so getting closer to the TX816 in terms of hardware is definitely a plausible thing for MiniDexed to do in my view :)

Kevin

**Reply by SetTheTone** [reply]
> ...or is what you _really_ would like to see a MiniDexed VST for the DAW? :)

No, because I often use my hardware live and try and not to have a laptop on stage, where I use a hardware sequencer like the Yamaha QY100 ;)

**Comment by diyelectromusic** [comment]
I do believe this build of MiniDexed will do this: https://github.com/diyelectromusic/MiniDexed/tree/QuadDAC8Chan.  NB: it needs the latest develop branch of circle...

You need to enable the following in your minidexed.ini file:
QuadDAC8Chan=1

I actually really like this - do let me know if you manage to give it a try - especially if you've got one of the aforementioned dac8x devices (I don't - I'm using four GY-PCM5102 modules!).

Kevin

**Reply by SetTheTone** [reply]
Hi Kevin

I have popped myself on the Hifiberry waiting list so hopefully soon I'll have the bits, I'm kind of a techy but new to all this Pi stuff so I'll try and get my head around it all but it sounds like I have a good point of call with you and probonopd

Milton

**Reply by diyelectromusic** [reply]
Now issued as a PR and there is a test build available.  See: https://github.com/probonopd/MiniDexed/pull/657

Kevin

**Comment by Banana71** [comment]
Thanks to @SetTheTone SetTheTone for reporting and thanks to Kevin @diyelectromusic for implementing it. I'm always surprised how quickly ideas are implemented, amazing.

I have a few questions:
- Would it work with just two GY-PCM5102 modules?
- Could an RPi 4 handle multiple GY-PCM5102 modules?

How I would imagine using the miniDexed and the option of multiple audio outputs:
- I would keep the PAN settings of each TG and add an additional parameter (Audio Out: 1 -4). This way, every imaginable scenario could be implemented.
- The internal reverb is permanently routed to Output 1. The remaining outputs could then be supplied with effects via the external mixer or DAW.

Examples:
Strings, Piano, Kick and Snare
TG1 Strings, PAN L, Audio Out 1, Reverb On
TG2 Strings, PAN R, Audio Out 1, Reverb On
TG3 Strings, PAN 32, Audio Out 1, Reverb On
TG4 Strings, PAN 96, Audio Out 1, Reverb On
TG5 Piano, PAN L, Audio Out 2
TG6 Piano, PAN R, Audio Out 2 
TG7 Kick, PAN L, Audio Out 3
TG8 Snare, PAN R, Audio Out 3
In this example you would have 2 stereo outputs and two individual outputs for the kick and snare.

**Reply by diyelectromusic** [reply]
Only the Pi 5 supports four I2S lanes via the new RP1 chip, so no, this is not possible on the Pi 4 in hardware (there might be software workarounds for another virtual I2S interface, but that is a major piece of bespoke work so really not worth the effort).

It could support just two modules, but the code is "all or nothing" at the moment and assumes either a single module or all four.

It could be made a lot more complex with possibilities as you suggest, but right now I've kept it simple.  I think the key reasoning being that if you want all TGs to have their own outputs then that is because you want hardware control over the mix alongside everything else.  Zynthian has a large range of complex MIDI routing options, processing chains and so on... and whilst powerful, I'm not sure it is the better for all that complexity, but then I'm hardly a "power user" of it atm :)

It might be worth implementing an audio out channel option in the performance subsystem though as you say - I'll have a ponder, but as I say for this initial look, I think simplicity is the way to go, so I'll leave that to a future change.

Detail: It's a bit more complicated with panning and effects as they require a software mixing stage which at present assumes a single stereo channel - so there would be a lot more to change compared to just ignoring the mixing and effects stage and dumping the raw samples out to the audio interface for external processing - which is what I've done here.

Kevin

**Reply by Banana71** [reply]
It's always a balancing act between "what is technically possible" and "what makes sense for the application". At the moment I'm not sure what I really want and need.

But even though the single output solution requires an RPi 5, it is the most cost-effective solution to get as close as possible to the TX816 model.

**Reply by probonopd** [reply]
For me _personally_, the sweet spot is to get close to the TX816 sound including commonly used 80s effects, mixed down to Stereo, using as minimalistic hardware as possible. But of course different people have different needs, and the flexibility is what sets a DIY project apart from an off-the-shelf product. We just need to balance configurability with simplicity vs. complexity.

**Reply by SetTheTone** [reply]
Maybe a simple configuration routine to allow adjustment for each day's requirements, that way you could re-configure on the fly, maybe with saved favourites, and an initial setup to select Pi type and installed memory.

Milton

**Reply by probonopd** [reply]
Luckily, Pi type and memory are autodetected.

**Comment by freddy-a** [comment]
This would be very welcome and useful. I'd be able to output different instances to different channels to my mixer, each with their own hardware effects. Would be great if for each pair of output channels you can choose if it's 2 mono channels or 1 stereo channel.
Like for example
TG1 E.Piano --> Stereo (ch. 1 & 2)
TG2 Slap Bass --> Mono (ch 3)
TG3 Drums --> Mono (ch 4)

And if I wanted the snare to have more reverb then the rest of the drums, I could add 
TG4 Drum --> Mono (ch 5) (with extra reverb in TG4 or via hardware)
And then not use the snare in TG3 but in TG4.

Very cool!
